fix fuzzy text on high resolution displays

ankye opened the pull request:
Re: cocos-creator/3d-tasks#

Changelog:
 * 
修复高分辨率文本显示模糊的问题
<!-- Note: Makes sure these boxes are checked before submitting your PR - thank you!

- [ ] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
- [ ] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.

- To official teams:
  - [ ] Check that your javascript is following our [style guide](https://github.com/cocos-creator/fireball/blob/dev/.github/CONTRIBUTING.md) and end files with a newline
  - [ ] Document new code with comments in source code based on [API Docs](https://github.com/cocos-creator/fireball#api-docs)
  - [ ] Make sure any runtime log information in `cc.log` , `cc.error` or `new Error('')` has been moved into `DebugInfos.js` with an ID, and use `cc.logID(id)` or `new Error(cc._getErrorID(id))` instead.

-->

soulsxh committed:
fix fuzzy text on high resolution displays

soulsxh committed:
fix fuzzy text on high resolution displays

shrinktofit committed:
Fix AnimationGraph.disconnect() (#9785)

* Fix AnimationGraph.disconnect()

* Update comment

PPpro committed:
fix rebase branch on interface check CI (#9786)

* fix rebase branch on interface check CI

* remove package-lock.json

* add running state

* Revert "add running state"

This reverts commit 49cb1689313f567598788d1516b5818f301a59ba.

hana-alice committed:
fix webgl2.0 context lost when launched (#9788)

PPpro committed:
fix the same name of branches for interface check (#9789)

* fix the same name of branches for interface check

* update

* fix rebase

jk20012001 committed:
Add scene panel component help url and item tooltips (#9783)

* 增加场景面板help和tooltips

* 增加场景面板help和tooltips

* update tooltips

* Apply suggestions from code review

Co-authored-by: yufang.wu <2435731537@qq.com>

* Apply suggestions from code review

Co-authored-by: yufang.wu <2435731537@qq.com>

Co-authored-by: yufang.wu <2435731537@qq.com>

troublemaker52025 committed:
fix shadowMap sampler accuracy loss for v3.4.0 (#9794)

pandamicro committed:
Upgrade internal asset meta (#9800)

yanOO1497 committed:
fix image save meta (#9799)

gameall3d committed:
fix __prefab copy (#9792)

known eslint issue in `instantiate.ts`

gameall3d committed:
fix prefabInstance creation under mounted children (#9804)

xianyinchen committed:
Fix particle display anomaly. (#9795)

SantyWang commented:
感谢贡献，3.4 已接近尾声，我们将在下个小版本合并 PR

LinYunMo committed:
fix bmfont line width bug (#9793)

* fix bmfont line width bug

* Fix CI

* fix CI

hana-alice committed:
fix crash on mtk chip (#9807)

hana-alice committed:
append effect.ts (#9808)

* fix crash on mtk chip

* append effect.ts

* append glsl

zhakesi committed:
fix natvie gc function (#9810)

jk20012001 committed:
Fix #10433: 打开safari webgl1 ddx ddy指令的扩展 (#9820)

SantyWang committed:
bump engine version (#9818)

jk20012001 committed:
update compiled shaders (#9822)

* update compiled shader

* Fix #10433: 打开safari webgl1 ddx ddy指令的扩展 (#9820)

LinYunMo committed:
remove DrawCall in DrawBatch (#9824)

pandamicro committed:
change default alpha blend params (#9784) (#9825)

Co-authored-by: zxx43 <zxx43@hotmail.com>

pandamicro committed:
Merge branch 'v3.4.0' into v3.4.1

# Conflicts:
#	cocos/core/builtin/effects.ts
#	cocos/core/builtin/shader-sources/glsl1.ts

PPpro committed:
compability for undefined game frame (#9830)

LinYunMo committed:
add profiler reset (#9806)

PPpro commented on cocos/2d/assembler/label/font-utils.ts:
```diff
@@ -36,6 +36,7 @@ import { BufferTextureCopy } from '../../../core/gfx';
 import { safeMeasureText, BASELINE_RATIO, MIDDLE_RATIO, getBaselineOffset } from '../../utils/text-utils';
 import { director, Director } from '../../../core/director';
 import { macro, warnID } from '../../../core';
+import { screenAdapter } from 'pal/screen-adapter';
```
请修复一下 ESLint 的报错，需要把 该 import 语句放在第一个 import 位置

PPpro commented on cocos/2d/assembler/label/font-utils.ts:
```diff
@@ -58,9 +59,13 @@ export class CanvasPool {
             data = {
                 canvas,
-                context,
+                context, 
```
请修复一下 ESLint 的报错，需要去掉行尾空格

PPpro commented on cocos/2d/assembler/label/font-utils.ts:
```diff
@@ -163,15 +168,20 @@ class LetterTexture {
-
-        if (this.canvas.width !== this.width) {
-            this.canvas.width = this.width;
+        const dpr = Math.ceil(screenAdapter.devicePixelRatio);
```
这里为什么需要做 ceil 操作呢

ankye replied:
devicePixelRatio有可能是小数，width和height要取整，缩放系数计算会有偏差，或者dpr取整，缩放也按dpr计算，按照dpr算高分屏最多也就持平正常效果，向上取整算出来清晰度能保证

SantyWang replied:
是否用 Math.round 比较合适呢？dpr 是 1.1 的时候，用 1 应该也是合适的吧

ankye replied:
之前做了一次测试，dpr2.25 cell和round区别不大，不太确定在canvas这块是否有太大性能影响

ankye replied:
对比dpr和label放大再缩小，清晰度差不多，但是放大看的话，放大再缩小噪点控制的更好，这个是引擎做了什么特殊处理吗？

ankye replied:
采用dpr后，再放大缩小，效果也有提升，所以当时就选了cell

PPpro replied:
我会比较偏向于使用 Math.round， 牺牲一些质量，
因为如果 dpr 2.25 ，用 2 倍跟 3 倍差异还是蛮大，3 倍的话，大部分情况是做无用功

@holycanvas @ankye 觉得呢

ankye replied:
我修改下

PPpro commented on cocos/2d/assembler/label/ttfUtils.ts:
```diff
@@ -40,6 +40,7 @@ import { legacyCC } from '../../../core/global-exports';
 import { assetManager } from '../../../core/asset-manager';
 import { dynamicAtlasManager } from '../../utils/dynamic-atlas/atlas-manager';
 import { BlendFactor } from '../../../core/gfx';
+import { screenAdapter } from 'pal/screen-adapter';
```
此处应该也需要置顶  import 语句

MSoft1115 committed:
fix terrain _updateHeight texcoord error (#9809)

zxx43 committed:
add burst restart (#9832)

ankye committed:
modify devicePixelRatio scale factor and fix eslint

ankye committed:
modify devicePixelRatio scale factor and fix eslint

ankye committed:
modify devicePixelRatio scale factor and fix eslint

ankye committed:
modify devicePixelRatio scale factor and fix eslint

ankye committed:
fix eslint error

ankye committed:
fix eslint error

pandamicro committed:
Merge pull request #9839 from cocos-creator/v3.4.0

Merge v3.4.0 official

SantyWang commented on cocos/2d/assembler/label/font-utils.ts:
```diff
@@ -163,15 +165,20 @@ class LetterTexture {
+        const w = this.width * dpr;
+        const h = this.height * dpr;
+        if (this.canvas.width !== w) {
+            this.canvas.style.width = `${this.width}px`;
```
这里为啥还要调整 style 中的属性呢，width 和 height 应该够了吧，style 中的属性在各个平台无法保证兼容性

ankye replied:
通过dpr设置width和height，整个canvas实际上是放大了dpr倍，通过style缩回canvas的实际显示大小，style不能保证兼容性这个我还没有注意，那有其他的替代方案吗？

ankye replied:
目前引擎都是按px走的，对接dpr回来就要做这次缩放处理，目前我能想到的是这样的处理过程，所以引入了style，如果有兼容性问题可以不合入，有解决方案再合，当前px的方式在引擎其他部分可能也有显示模糊或者其他的问题，是否需要通盘考虑？

ankye replied:
https://www.w3schools.com/cssref/pr_dim_height.asp  我查了一下，width和height是全浏览器都支持的，这个兼容性指？

SantyWang replied:
主要是各种小游戏平台，如果不设置 style 中的宽高，会有什么问题么

ankye replied:
理论上文字放大了，引擎取到的纹理就可能只有一部分了，没实验过

ankye replied:
我试了一下，android和pc都没有问题

ankye replied:
看了一下label，直接拿context处理的，context做了scale，去掉目前看没有问题，可以请写这块的童鞋review一下

ankye replied:
刚在win上测试了一下，有概率出现文字变大的问题，我需要一些时间定位

ankye replied:
@holycanvas @PPpro 先不合吧，目前发现几个问题，1. bitmap和char模式没有处理，2，label的overflow没有处理，3.atlas texture回来的计算没有考虑

ankye replied:
这两天在改bitmap和char模式，通过dpr计算，清晰度上来了，但是对纹理的占用也增加了，char模式下1024x1024和bitmap下2048x2048的占用都放不了太多文字了，尤其是3倍的情况，一个fontsize=40的字，占用 120x120的文字尺寸 @PPpro @holycanvas

ankye replied:
我修改完了，是否需要锁2倍dpr，2倍字体清晰度感觉也够了，eslint是否有设置方式，我现在vscode本地校验都过了，提交上来就跪，重新提pr，好像又给开了一个新的，有点没玩懂 @holycanvas @PPpro

troublemaker52025 committed:
updateLightingmap sync to native-engine for v3.4.1 (#9828)

* updateLightingmap sync to native-engine for v3.4.1

* update

* merge JSB binding

xubing0906 committed:
repair UseHDR button& Skybox HDR status (#9843)

Co-authored-by: xubingbing <bingbing.xu@cocos.com>

Zhmz committed:
add decorator “serializable” (#9848)

Co-authored-by: zhangmingzhen <mingzhen.zhang@cocos.com>

PPpro committed:
resctrict DPR on Web platform (#9850)

zxx43 committed:
Improve gpu particle performance (#9847)

* improve gpu particle performance

* fix eslint error

YunHsiao committed:
disable multi_draw extension on mobile (#9835)

yiwenxue committed:
Add extra steps to the destroy functions to update the cached resources. (#9854)

* Add extra steps to the destroy functions to update the cached resources.
This will make sure that the resources in device.stateCache is valid.

* unbind a resource before destroying it

* remove unnecessary non-null assertion

* update the state after the destruction to make the process robust

PPpro committed:
cache systemInfo on Cocos Play platform (#9856)

* cache systemInfo on Cocos Play platform

* resize swapChain

jk20012001 committed:
Fix skeleton model receive shadow macro error and toon shadow display errors (#9836)

* 修复骨骼模型接受阴影的开关无效 及 toon shadow计算的错误

* toon精确判断背光面及强制所有面接受阴影

* toon使用参数来调节接受阴影的区域

troublemaker52025 committed:
fix shadow queue perfprmance for v3.4.1 (#9862)

* fix shadow queue perfprmance for v3.4.1

* update

Zhmz committed:
Mask not work with spine v3.4.1 (#9864)

* add condition spine preRenderComponent validation before autoMergeBatches

(cherry picked from commit 04fcebde46ec264c0f09e6afdcba54ae6935a1fc)

* add explanations

(cherry picked from commit 72f8439cfdeda28d4f9b944b4f3d5640bbd05fb3)

Co-authored-by: zhangmingzhen <mingzhen.zhang@cocos.com>

jk20012001 committed:
Fix #10538: Android+WebGL PS struct variables has mediump precision (#9863)

* Fix #10538: Android+WebGL PS struct variables has mediump precision

* update compiled shaders

* review modification

* update compiled shaders

jk20012001 committed:
fix define syntax error (#9870)

* fix define syntax error

* fix toon shader syntax error

LinYunMo committed:
Revert "Add extra steps to the destroy functions to update the cached resources. (#9854)" (#9873)

This reverts commit acb3fce6074d43f5fb40172fa765036bf20e22f8.

Co-authored-by: Yiwen <15225434259xue@gmail.com>

minggo committed:
detach from scene before destroying Model/Light (#9869)

PPpro committed:
fix set frame rate on runtime (#9874)

ankye committed:
Merge branch 'cocos-creator:v3.4.0' into v3.4.0

ankye committed:
fix fuzzy text on high resolution displays

ankye committed:
fix fuzzy text on high resolution displays

fish9167 committed:
Add GitHub Action (#9872)

* Add GitHub Action

Add GitHub Action： run test cases

* Update run_test_cases.yml

* Update run_test_cases.yml

* Update run_test_cases.yml

* Update run_test_cases.yml

* Update run_test_cases.yml

* Update run_test_cases.yml

* Update run_test_cases.yml

* Update run_test_cases.yml

retry github action

* Update run_test_cases.yml

updata npm

* Update run_test_cases.yml

* Update run_test_cases.yml

* Update run_test_cases.yml

* Update run_test_cases.yml

retry github action

* Update run_test_cases.yml

YunHsiao committed:
material.copy accepts overrides (#9881)

Zhmz committed:
Label string setter optimization (#9880)

* remove rebundant onDisable

* update label.string.setter and add unit test case

Co-authored-by: zhangmingzhen <mingzhen.zhang@cocos.com>

yanjifa committed:
fix Color.fromHex Bug (#9867)

arsen2010 committed:
change ts template (#9883)

pandamicro commented:
Please rebase to latest v3.4.1, the commit history shouldn't contain any other commits, v3.4.0 branch should have been merged to v3.4.1

troublemaker52025 committed:
Optimize fixedArea mode culling process for v3.4.1 (#9878)

ankye committed:
Merge branch 'v3.4.0' of github.com:ankye/engine into v3.4.0
